### PR TITLE
[recipe] AMI Zipformer transducer

### PR DESCRIPTION
**Pull request by desh2608**
This is a Zipformer based pruned transducer recipe for AMI dataset. We train a single model that can work well on all the conditions (IHM, SDM, enhanced MDM). WER results using modified beam search (`--epoch 14 --avg 8 --use-averaged-model True`):

| Evaluation set           | dev WER    | test WER |
|--------------------------|------------|---------|
| IHM                      |  18.92  | 17.40 |
| SDM                      |  31.25  | 32.21 |
| MDM (GSS-enhanced)       |  21.67  | 22.43 |

The GSS enhancement is based on [this package](https://github.com/desh2608/gss).

### CODE REVIEW

**Comment by csukuangfj** [comment]
So the results are better than that of
- kaldi
  - https://github.com/kaldi-asr/kaldi/blob/master/egs/ami/s5/RESULTS_ihm
  - https://github.com/kaldi-asr/kaldi/blob/master/egs/ami/s5/RESULTS_sdm
  - https://github.com/kaldi-asr/kaldi/blob/master/egs/ami/s5/RESULTS_mdm

and
- espnet
  - https://github.com/espnet/espnet/blob/master/egs/ami/asr1/RESULTS_ihm.md

---

https://github.com/espnet/espnet/tree/master/egs2/ami/asr1
has a lower WER in ihm dev/test dataset.

File: egs/ami/ASR/RESULTS.md [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+contains only an embedding layer, a Conv1d (with kernel size 2) and a linear
+layer (to transform tensor dim).
+
+All the results below are using a single model that is trained by combining the following
+data: IHM, IHM+reverb, SDM, and GSS-enhanced MDM. Speed perturbation and MUSAN noise
+augmentation are applied on top of the pooled data.
+
+**WERs for IHM:**
+
+|                           | dev | test | comment                                  |
```
Comment (csukuangfj) [inline egs/ami/ASR/RESULTS.md:19]: Do you have results for `greedy search`?

File: egs/ami/ASR/pruned_transducer_stateless7/decode.py [hunk 2]
```diff
@@ -0,0 +1,681 @@
...
+        type=float,
+        default=0.5,
+        help="""Scale applied to lattice scores when computing nbest paths.
+        Used only when the decoding method is fast_beam_search_nbest,
+        fast_beam_search_nbest_LG, and fast_beam_search_nbest_oracle""",
+    )
+
+    add_model_arguments(parser)
+
+    return parser
```
Comment (csukuangfj) [inline egs/ami/ASR/pruned_transducer_stateless7/decode.py:236]: Could you use the decode.py from librispeech?

https://github.com/k2-fsa/icefall/blob/500792d0f1d8da67c1bd70b3066c78cf36811ce3/egs/librispeech/ASR/pruned_transducer_stateless7/decode.py#L173
the above option is crucial for getting better results in decoding.

File: egs/ami/ASR/pruned_transducer_stateless7/asr_datamodule.py [hunk 3]
```diff
@@ -0,0 +1,430 @@
...
+    ) -> DataLoader:
+        """
+        Args:
+          cuts_train:
+            CutSet for training.
+          sampler_state_dict:
+            The state dict for the training sampler.
+        """
+        logging.info("About to get Musan cuts")
+        cuts_musan = load_manifest(self.args.manifest_dir / "musan_cuts.jsonl.gz")
```
Comment (csukuangfj) [inline egs/ami/ASR/pruned_transducer_stateless7/asr_datamodule.py:211]: Shall we move it after line 214?

**Comment by desh2608** [comment]
> So the results are better than that of
> 
> * kaldi
>   
>   * https://github.com/kaldi-asr/kaldi/blob/master/egs/ami/s5/RESULTS_ihm
>   * https://github.com/kaldi-asr/kaldi/blob/master/egs/ami/s5/RESULTS_sdm
>   * https://github.com/kaldi-asr/kaldi/blob/master/egs/ami/s5/RESULTS_mdm
> 
> and
> 
> * espnet
>   
>   * https://github.com/espnet/espnet/blob/master/egs/ami/asr1/RESULTS_ihm.md
> 
> https://github.com/espnet/espnet/tree/master/egs2/ami/asr1 has a lower WER in ihm dev/test dataset.

Yes, that's right. It's not directly comparable though since all of those train for specific conditions, e.g., for IHM, or for SDM. I was going for a more general-purpose model. The ESPNet2 model without LM is ~1% better.

File: egs/ami/ASR/RESULTS.md (hunk 1 above)
Comment (desh2608) [reply egs/ami/ASR/RESULTS.md:19]: Will add it.

File: egs/ami/ASR/pruned_transducer_stateless7/decode.py (hunk 2 above)
Comment (desh2608) [reply egs/ami/ASR/pruned_transducer_stateless7/decode.py:236]: I see. Let me try decoding with this option.

Comment (desh2608) [reply egs/ami/ASR/pruned_transducer_stateless7/decode.py:236]: Using this gives me slightly worse WERs.

Comment (csukuangfj) [reply egs/ami/ASR/pruned_transducer_stateless7/decode.py:253]: What `--avg` value are you using?

Comment (desh2608) [reply egs/ami/ASR/pruned_transducer_stateless7/decode.py:253]: I used --iter 105000, --avg 9.

Comment (csukuangfj) [reply egs/ami/ASR/pruned_transducer_stateless7/decode.py:253]: Have you tried other values of `--avg` instead of `9`. Also, have you tried other combinations of `--epoch` and `--avg`?

Comment (csukuangfj) [reply egs/ami/ASR/pruned_transducer_stateless7/decode.py:253]: Ok, you are using `--iter`. Maybe you can also try `--epoch`.

Since you only train 15 epochs, `--avg 9` may be too large.

Comment (csukuangfj) [reply egs/ami/ASR/pruned_transducer_stateless7/decode.py:253]: > Using this gives me slightly worse WERs.

This is the first time that I have seen that it causes a worse WER, I think.

Comment (desh2608) [reply egs/ami/ASR/pruned_transducer_stateless7/decode.py:253]: I will try more combinations and report here today. I tried that one because it was giving me the best WERs without the --use-averaged-model, but perhaps it may not be optimal when used with this option.

Comment (desh2608) [reply egs/ami/ASR/pruned_transducer_stateless7/decode.py:253]: I tried using different number of epoch checkpoints for averaging, and here are the results using fast beam search. The first row is without `--use-averaged-model`. It seems averaging 7, 8, or 9 of the last epochs gives improvements.

| **Decoding method**                   | **Dev IHM** | **Test IHM** | **Dev SDM** | **Test SDM** | **Dev GSS** | **Test GSS** |
|---------------------------------------|-------------|--------------|-------------|--------------|-------------|--------------|
| FBS (ckpt=105000, avg=10)             |    19.46    |     18.35    |    31.14    |     32.52    |    22.45    |     23.38    |
| FBS (epoch=14, avg=5, use-avg-model)  |    19.75    |     18.63    |    31.27    |     32.26    |     22.7    |     23.58    |
| FBS (epoch=14, avg=6, use-avg-model)  |    19.52    |     18.43    |    31.13    |     32.07    |    22.47    |     23.24    |
| FBS (epoch=14, avg=7, use-avg-model)  |    **19.42**    |     18.17    |   **31.01**    |     **32.04**    |     22.3    |     22.98    |
| FBS (epoch=14, avg=8, use-avg-model)  |    19.44    |     18.04    |    31.11    |     32.1     |    **22.21**    |     22.83    |
| FBS (epoch=14, avg=9, use-avg-model)  |    19.49    |     **17.99**    |    31.34    |     32.35    |    22.22    |     **22.78**    |
| FBS (epoch=14, avg=10, use-avg-model) |    19.55    |     18.04    |    31.85    |     32.88    |    22.41    |     23.03    |

Comment (desh2608) [reply egs/ami/ASR/pruned_transducer_stateless7/decode.py:253]: I will update the results to use `--epoch 14 --avg 8 --use-averaged-model` since it seems to provide consistent improvements across all sets.

Comment (desh2608) [reply egs/ami/ASR/pruned_transducer_stateless7/decode.py:253]: Done.

**Comment by csukuangfj** [comment]
Please mark this PR with the label `ready` when you think it is ready to merge. 
Thanks!

**Comment by desh2608** [comment]
> Please mark this PR with the label `ready` when you think it is ready to merge. Thanks!

Perhaps we can get this merged if there are no further changes required?

**Comment by csukuangfj** [comment]
> > Please mark this PR with the label `ready` when you think it is ready to merge. Thanks!
> 
> Perhaps we can get this merged if there are no further changes required?

Thanks!

**Comment by csukuangfj** [comment]
@desh2608 
Could you also upload `tokens.txt` to https://huggingface.co/desh2608/icefall-asr-ami-pruned-transducer-stateless7/tree/main/data/lang_bpe_500

Also, could you upload some test waves to the above huggingface repo, just like the following one?
https://huggingface.co/csukuangfj/icefall-asr-librispeech-pruned-transducer-stateless5-2022-05-13/tree/main/test_wavs

**Comment by desh2608** [comment]
> @desh2608 Could you also upload `tokens.txt` to https://huggingface.co/desh2608/icefall-asr-ami-pruned-transducer-stateless7/tree/main/data/lang_bpe_500
> 
> Also, could you upload some test waves to the above huggingface repo, just like the following one? https://huggingface.co/csukuangfj/icefall-asr-librispeech-pruned-transducer-stateless5-2022-05-13/tree/main/test_wavs

Done!
